### DISCUSSION TITLE
RPC/HTTP API?

### DISCUSSION THREAD

**Discussion by ia74** [rekkyrosso/ampcast#22] [Ideas]
Hey there! I've been using ampcast for a while now, but I think that the app should have some sort of API that can for example query the currently playing song or control the player itself (play/pause/next/back/vol).
I'd like to make a [plugin for my project Freedeck](https://github.com/Freedeck/freedeck)

**Comment by rekkyrosso** [comment]
I'm not sure how that might work. Can you sketch out a basic example?

**Reply by ia74** [reply]
My initial thought was to have a WebSocket server that can control the client, like this:
1. WS opens on /api on the [server](https://github.com/rekkyrosso/ampcast/blob/main/server.js)
2. Ampcast connects to the WebSocket server, and the server remembers that its the app for events
3. A client can connect to http://localhost:8000/api
4. Maybe here there would be server-side authentication with a sort of global password
5. Periodically Ampcast will send down the player state (time, title, etc) and the server will broadcast that to all connected clients
6. Running a command will get forwarded to Ampcast to execute it
(Example) sending the packet `{event:"print",args:["a"]}` will get forwarded (Client->Server->Ampcast) and Ampcast will log "a"

**Reply by rekkyrosso** [reply]
Ampcast is written as a ~~dumb~~ pure client. The server component is just to serve the application files.

If your use case is remote control for play/pause/etc then I'm not against the idea of using your phone like that. Just not sure how.
